feMerge misplaces the HTML element when merging the same feMergeNode multiple times

shallawa opened the pull request:
#### e2873429150ba8df42e6acf17a0ff6ac64f1a69f
<pre>
feMerge misplaces the HTML element when merging the same feMergeNode multiple times
<a href="https://bugs.webkit.org/show_bug.cgi?id=291919">https://bugs.webkit.org/show_bug.cgi?id=291919</a>
<a href="https://rdar.apple.com/149431216">rdar://149431216</a>

Reviewed by Simon Fraser.

Two bugs need to be fixed to get this scenario working:

1. Building SVGFilterPrimitiveGraph:

There are two graphs for SVGFiler. One is for the effects and the other is for
the primitives. The main difference between these two graphs is the effect graph
has entries for SourceGraphics and SourceAlpha. The primitive graph can&apos;t have
entries for these predefined effects. But there should not be an error when
handling them.

When looking for a named effect, we fall back to the lastEffect when there is no
matched named effect and it is not predefined effect. Since there is no predefined
primitives, getNamedNode() was falling back to the lastNode when the name is
SourceGraphis or SourceAlpha. This was building a wrong primitives graph.

To make fixing this simple, SVGFilterPrimitiveGraph and SVGFilterEffectGraph
should be split and should be made sub-classes of SVGFilterGraph.
Only SVGFilterEffectGraph will handle the predefined effects.

2. The outsets of the sibling primitives should be their maximum:

SVGFilter::calculateOutsets() will push the outsets of the current primitive
associated with the its level into a stack. When the graph enumeration goes up
one level, all the outsets on the lower level will be popped out of the stack
and their maximum will expand the outsets of the top of outsets stack.

* LayoutTests/svg/filters/feMerge-nodes-unite-outsets-expected.html: Added.
* LayoutTests/svg/filters/feMerge-nodes-unite-outsets.html: Added.
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/RectEdges.h:
(WebCore::blend):
(WebCore::max):
* Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
(WebCore::buildFilterEffectGraph):
(WebCore::SVGFilter::buildExpression):
(WebCore::buildFilterPrimitiveGraph):
(WebCore::SVGFilter::isIdentity):
(WebCore::SVGFilter::calculateOutsets):
(WebCore::buildFilterEffectsGraph): Deleted.
(WebCore::buildFilterPrimitivesGraph): Deleted.
* Source/WebCore/svg/graphics/filters/SVGFilterEffectGraph.h: Added.
(WebCore::SVGFilterEffectGraph::SVGFilterEffectGraph):
(WebCore::SVGFilterEffectGraph::sourceGraphic const):
(WebCore::SVGFilterEffectGraph::sourceAlpha const):
* Source/WebCore/svg/graphics/filters/SVGFilterGraph.h:
(WebCore::SVGFilterGraph::isSourceName):
(WebCore::SVGFilterGraph::SVGFilterGraph): Deleted.
(WebCore::SVGFilterGraph::sourceGraphic const): Deleted.
(WebCore::SVGFilterGraph::sourceAlpha const): Deleted.
(WebCore::SVGFilterGraph::addNamedNode): Deleted.
(WebCore::SVGFilterGraph::getNamedNode const): Deleted.
* Source/WebCore/svg/graphics/filters/SVGFilterPrimitiveGraph.h: Added.

Canonical link: <a href="https://commits.webkit.org/294088@main">https://commits.webkit.org/294088@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d864b4d80d3099ae5bda14f922538c4fc0b12f21

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100792 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20444 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10743 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/105929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/51381 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102833 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20753 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28918 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/105929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/51381 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103799 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/15951 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/91040 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/105929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/15761 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9049 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/50756 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/85662 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9123 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108284 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27910 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/20502 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/85713 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28273 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87242 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/85256 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21694 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29962 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/7690 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/21914 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27845 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/33101 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27656 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30974 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29214 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/RectEdges.h:
```diff
@@ -151,6 +151,8 @@ template<typename T> class RectEdges {
 
     bool operator==(const RectEdges<T>&) const = default;
 
+    void unite(const RectEdges<T>&);
```
This is a little odd to have here because type T might not be something you can call `max()` on.

shallawa replied:
Yes but:

1. For `RectEdges`, we define the `operator+`. This operator would not compile for `RectEdges<LayoutUnit>` if `operator+` was not defined `LayoutUnit`.
2. This is a template function. So the source will compile for `RectEdges<X>` even if type `X` does not define the comparison operators. The compiler will complain only if you call `maximum(x1, x2)`

smfr commented on Source/WebCore/platform/RectEdges.h:
```diff
@@ -173,6 +175,23 @@ inline RectEdges<T>& operator+=(RectEdges<T>& a, const RectEdges<T>& b)
+        return;
+    }
+
+    setTop(std::max(top(), other.top()));
```
This seems to be wrong. Top/left should use `min()`.

shallawa replied:
No I do not think so. The top and left of `RectEdges` are directionless lengths. For example, when a `FloatBoxExtent` is added to a `FloatRect` it expands the `FloatRect`:
```
    void expand(const FloatBoxExtent& box)
    {
        m_location.move(-box.left(), -box.top());
        m_size.expand(box.left() + box.right(), box.top() + box.bottom());
    }
```

smfr commented on Source/WebCore/platform/RectEdges.h:
```diff
@@ -173,6 +175,23 @@ inline RectEdges<T>& operator+=(RectEdges<T>& a, const RectEdges<T>& b)
+    }
+
+    setTop(std::max(top(), other.top()));
+    setTop(std::max(right(), other.right()));
```
setRight?

smfr replied:
Thinking more, `unite()` on RectEdges makes no sense. It's not a Rect, so you can't unite it, and unite probably doesn't make any sense for many values of`T`.

shallawa replied:
I renamed it '`max` since it gets the maximum for all the four sides of `RectEdges`.

smfr commented on Source/WebCore/svg/graphics/filters/SVGFilterPrimitiveGraph.h:
```diff
@@ -0,0 +1,60 @@
+
+class SVGFilterPrimitiveStandardAttributes;
+
+class SVGFilterPrimitiveGraph : public SVGFilterGraph<SVGFilterPrimitiveStandardAttributes> {
```
`final`

smfr commented on Source/WebCore/svg/graphics/filters/SVGFilterEffectGraph.h:
```diff
@@ -0,0 +1,91 @@
+
+class FilterEffect;
+
+class SVGFilterEffectGraph : public SVGFilterGraph<FilterEffect> {
```
`final`

shallawa committed:
feMerge misplaces the HTML element when merging the same feMergeNode multiple times
https://bugs.webkit.org/show_bug.cgi?id=291919
rdar://149431216

Reviewed by Simon Fraser.

Two bugs need to be fixed to get this scenario working:

1. Building SVGFilterPrimitiveGraph:

There are two graphs for SVGFiler. One is for the effects and the other is for
the primitives. The main difference between these two graphs is the effect graph
has entries for SourceGraphics and SourceAlpha. The primitive graph can't have
entries for these predefined effects. But there should not be an error when
handling them.

When looking for a named effect, we fall back to the lastEffect when there is no
matched named effect and it is not predefined effect. Since there is no predefined
primitives, getNamedNode() was falling back to the lastNode when the name is
SourceGraphis or SourceAlpha. This was building a wrong primitives graph.

To make fixing this simple, SVGFilterPrimitiveGraph and SVGFilterEffectGraph
should be split and should be made sub-classes of SVGFilterGraph.
Only SVGFilterEffectGraph will handle the predefined effects.

2. The outsets of the sibling primitives should be their maximum:

SVGFilter::calculateOutsets() will push the outsets of the current primitive
associated with the its level into a stack. When the graph enumeration goes up
one level, all the outsets on the lower level will be popped out of the stack
and their maximum will expand the outsets of the top of outsets stack.

* LayoutTests/svg/filters/feMerge-nodes-unite-outsets-expected.html: Added.
* LayoutTests/svg/filters/feMerge-nodes-unite-outsets.html: Added.
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/RectEdges.h:
(WebCore::blend):
(WebCore::max):
* Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
(WebCore::buildFilterEffectGraph):
(WebCore::SVGFilter::buildExpression):
(WebCore::buildFilterPrimitiveGraph):
(WebCore::SVGFilter::isIdentity):
(WebCore::SVGFilter::calculateOutsets):
(WebCore::buildFilterEffectsGraph): Deleted.
(WebCore::buildFilterPrimitivesGraph): Deleted.
* Source/WebCore/svg/graphics/filters/SVGFilterEffectGraph.h: Added.
(WebCore::SVGFilterEffectGraph::SVGFilterEffectGraph):
(WebCore::SVGFilterEffectGraph::sourceGraphic const):
(WebCore::SVGFilterEffectGraph::sourceAlpha const):
* Source/WebCore/svg/graphics/filters/SVGFilterGraph.h:
(WebCore::SVGFilterGraph::isSourceName):
(WebCore::SVGFilterGraph::SVGFilterGraph): Deleted.
(WebCore::SVGFilterGraph::sourceGraphic const): Deleted.
(WebCore::SVGFilterGraph::sourceAlpha const): Deleted.
(WebCore::SVGFilterGraph::addNamedNode): Deleted.
(WebCore::SVGFilterGraph::getNamedNode const): Deleted.
* Source/WebCore/svg/graphics/filters/SVGFilterPrimitiveGraph.h: Added.

Canonical link: https://commits.webkit.org/294088@main
